### PR TITLE
Re-evaluate clang tidy

### PR DESCRIPTION
**Pull request by mwaxmonsky**
Adds a `clang-tidy` workflow to static analyze most of the micm source code with a few exceptions noted in the filters in the workflow file.

Mostly required adding includes that were missing to compile the file in isolation and one template call fix in `cuda_linear_solver.hpp`.

Currently only uses the `bugprone-*` check for proof of concept but will open a discussion with documentation and use cases for other checks that can be turned on later.

Manually configures the configuration at the command line but will integrate with and further modify the `.clang-tidy` config file in #489 at a later date for consistency.

### CODE REVIEW

**Review by K20shores — APPROVED** [review]
love it

**Comment by K20shores** [comment]
@mwaxmonsky would you please file an issue detailing anything that's outstanding (like using or removing the cmake clang tidy stuff, and integrating your checks with the clang tidy file)?

**Comment by mwaxmonsky** [comment]
> @mwaxmonsky would you please file an issue detailing anything that's outstanding (like using or removing the cmake clang tidy stuff, and integrating your checks with the clang tidy file)?

Yup!  Started working on it this morning and will push up a generic issue similar to the GPU resident checklist issue with all of the issues linked.

File: include/micm/solver/cuda_linear_solver.cuh [hunk 1]
```diff
@@ -4,6 +4,9 @@
  */
 #pragma once
 
+
+#include <chrono>
```
Comment (sjsprecious) [inline include/micm/solver/cuda_linear_solver.cuh:8]: there is no timing measurement in this file so I don't think we need this header file?

File: include/micm/solver/cuda_linear_solver.hpp [hunk 2]
```diff
@@ -86,7 +86,7 @@ namespace micm
         const SparseMatrixPolicy<T>& L,
         const SparseMatrixPolicy<T>& U) const
     {
-      LinearSolver<T, SparseMatrixPolicy, LuDecompositionPolicy>::Solve<MatrixPolicy>(b, x, L, U);
+      LinearSolver<T, SparseMatrixPolicy, LuDecompositionPolicy>:: template Solve<MatrixPolicy>(b, x, L, U);
```
Comment (sjsprecious) [inline include/micm/solver/cuda_linear_solver.hpp:89]: Interesting. Why we need the `template` keyword here but not in other CUDA c++ files?

File: include/micm/util/sparse_matrix_vector_ordering.hpp [hunk 3]
```diff
@@ -4,9 +4,11 @@
  */
 #pragma once
 
+#include <functional>
```
Comment (boulderdaze) [inline include/micm/util/sparse_matrix_vector_ordering.hpp:7]: Doesn't look like this header is needed, could you confirm it?

File: include/micm/util/sparse_matrix_vector_ordering.hpp [hunk 4]
```diff
@@ -4,9 +4,11 @@
...
 #include <micm/util/matrix_error.hpp>
 #include <micm/util/vector_matrix.hpp>
 
+#include "sparse_matrix.hpp"
```
Comment (boulderdaze) [inline include/micm/util/sparse_matrix_vector_ordering.hpp:11]: For consistency in including project header files, could you change this?
```suggestion
#include <micm/util/sparse_matrix.hpp>
```

File: include/micm/util/vector_matrix.hpp [hunk 5]
```diff
@@ -10,6 +10,7 @@
 #include <algorithm>
 #include <cassert>
 #include <cmath>
+#include <stdexcept>
```
Comment (boulderdaze) [inline include/micm/util/vector_matrix.hpp:13]: Doesn't look like this header is needed, could you confirm it?

File: test/integration/analytical_policy.hpp [hunk 6]
```diff
@@ -19,6 +19,7 @@
 #include <iostream>
 #include <utility>
 #include <vector>
+#include <gtest/gtest.h>
```
Comment (boulderdaze) [inline test/integration/analytical_policy.hpp:22]: Looks like a duplicate with `line 15`. I believe clang-format will clean up though

**Review by boulderdaze — CHANGES_REQUESTED** [review]
Looks good! Just a couple minor change requests. I am curious how the header files are added. Was it an automatic process that clang is involved?

File: test/integration/analytical_policy.hpp (hunk 6 above)
Comment (mwaxmonsky) [reply test/integration/analytical_policy.hpp]: Good catch!  Removed!

File: include/micm/util/sparse_matrix_vector_ordering.hpp (hunk 4 above)
Comment (mwaxmonsky) [reply include/micm/util/sparse_matrix_vector_ordering.hpp]: Updated!

File: include/micm/util/cuda_dense_matrix.hpp [hunk 7]
```diff
@@ -12,6 +12,7 @@
 #include <cuda_runtime.h>
 
 #include <type_traits>
+#include <iostream>
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply include/micm/util/cuda_dense_matrix.hpp]: Nope!  Removed!

File: include/micm/solver/cuda_linear_solver.cuh (hunk 1 above)
Comment (mwaxmonsky) [reply include/micm/solver/cuda_linear_solver.cuh]: Nope!  Removed!

**Review by mattldawson — APPROVED** [review]
nice!

File: include/micm/solver/cuda_linear_solver.hpp (hunk 2 above)
Comment (mwaxmonsky) [reply include/micm/solver/cuda_linear_solver.hpp:89]: I need to dig more into this myself but this was the error (https://github.com/NCAR/micm/actions/runs/9001800054/job/24728778386#step:4:345) and an explanation behind this was https://stackoverflow.com/a/34732843 which I need to dig more into myself.

**Review by sjsprecious — APPROVED** [review]

**Review by boulderdaze — APPROVED** [review]
Thank you for addressing them!

File: test/unit/process/test_cuda_process_set.cpp [hunk 8]
```diff
@@ -14,6 +14,7 @@
 #include <iostream>
 #include <random>
 #include <vector>
+#include <algorithm>
```
[the remark this answers is not in the document]
Comment (mwaxmonsky) [reply test/unit/process/test_cuda_process_set.cpp]: Nope!  We were using it previously when we were calling `std::generate` but we migrated to using `std::ranges::generate` using the `ranges` library so removed!

**Pull request merged**